refactor(chat): Remove parenthesis wrapping of other files when calling `selectedCodePromptWithExtraFiles`

ichim-david opened the pull request:
refactor(chat): Remove file mention duplication for default commands.
Based on the comments:
https://github.com/sourcegraph/cody/pull/6910#pullrequestreview-2588426217
 https://github.com/sourcegraph/cody/pull/6910#issuecomment-2638393115

I'm opening this pull request in order to get rid of the extra setup parentheses between the primary mention and the other mentions.
For extra context on reason see this comment https://github.com/sourcegraph/cody/pull/6910#issuecomment-2629475322

I am also removing the duplication of context mention from the default commands because these commands have been duplicated as default prompts and they do not duplicate the context mention leaving just the context mention with the file range.

## Test plan
1. Run default commands such as explain code and notice the lack of context mention duplication
2. Run a prompt that uses a cody-context such as cody://current-tabs and notice that the tabs following the first tab are no longer wrapped with parenthesis
<!-- Required. See https://docs-legacy.sourcegraph.com/dev/background-information/testing_principles. -->

ichim-david commented:
@vovakulikov I'm going to need some help with the tests that fail because the polly recordings will need to be updated.
Can you please handle this as part of the code review?

dominiccooney commented on vscode/src/commands/execute/explain.ts:
```diff
@@ -45,7 +45,7 @@ export async function explainCommand(
     prompt = prompt.replaceAll(
         'the selected code',
-        selectedCodePromptWithExtraFiles(contextItems[0], contextItems.slice(1))
+        selectedCodePromptWithExtraFiles(contextItems[0], [])
```
I'm unsure about this (and the same changes in doc, etc.)

The intent here is to give the rest of the file because it is probably relevant for putting the selected code in context. With Agentic context it doesn't matter, but without it you might get a worse result. If the prompting is a problem, fine, let's fix the prompting but I'm not sure dropping the file mention is the way to go.

ichim-david replied:
@dominiccooney just because the second parameter is not passed doesn't mean that Cody doesn't pass the whole file as context and not just the range.
Here is a video where I show this:
https://www.youtube.com/watch?v=StQx8PQrCEM

The prompt indeed should be modified to also pass the entire file in order for the LLM to get a bit more context, this feels to me like another issue though.

Please take a look as I think that this addresses your concern, if not let me know.

ichim-david replied:
@dominiccooney if you look at the return statement the chat request receives the entire contextItems not just the first element. https://github.com/sourcegraph/cody/blob/7401a1457d22101e9af61350e671159e269528b3/vscode/src/commands/execute/explain.ts#L53

I'm only removing the extra display of the file in the chat window to avoid having  example.ts:10-15 example.ts show up since we removed the parenthesis.

if you look at my debug you can see that contextItems contains both and we return that so there is no worry of users receiving less context, we just clean up the interface for them.

![Image 26 02 2025 at 08 35](https://github.com/user-attachments/assets/50b6c309-22cc-4fdb-be92-2487ca4489cd)

ichim-david replied:
@dominiccooney  If you think that it's still valuable to show to the user what he receives than I add the second parameter back and it looks like this
![Image 26 02 2025 at 08 45](https://github.com/user-attachments/assets/1d06e020-4457-4c26-adec-9c99ccb96624)

dominiccooney reviewed (changes requested):
I'm unsure about this. What we probably need is a better prompt which asks to explain/document/etc. the <selected code> and then mentions the rest of the file as context later in the request.

dominiccooney commented on vscode/src/commands/execute/index.ts:
```diff
@@ -76,7 +76,10 @@ export function selectedCodePromptWithExtraFiles(
-    return ps`${primaryMention}${
-        otherMentions.length > 0 ? ps` ( ${PromptString.join(otherMentions, ps` `)} )` : ''
-    }`
+    if (otherMentions.length === 0) {
+        return primaryMention
+    }
+
+    const joinedMentions = PromptString.join(otherMentions, ps` `)
+    return ps`${primaryMention} ${joinedMentions}`
```
Being more succinct here will be more readable:

```suggestion
    return PromptString.join([primaryMention, ...otherMentions], ps` `)
```

dominiccooney commented on vscode/src/commands/execute/smell.ts:
```diff
@@ -45,7 +45,7 @@ async function smellCommand(
     prompt = prompt.replaceAll(
         'the selected code',
-        selectedCodePromptWithExtraFiles(contextItems[0], contextItems.slice(1))
+        selectedCodePromptWithExtraFiles(contextItems[0], [])
```
This is the problem. contextItems[1] contains the whole file and you are dropping it here.

dominiccooney commented on vscode/src/commands/execute/test-chat.ts:
```diff
@@ -50,7 +50,7 @@ async function unitTestCommand(
                 prompt = prompt.replaceAll(
                     'the shared code',
-                    selectedCodePromptWithExtraFiles(sharedContext[0], sharedContext.slice(1))
+                    selectedCodePromptWithExtraFiles(sharedContext[0], [])
```
...same here.

dominiccooney reviewed (changes requested):
More feedback inline.

I think you should not drop those extra context items from smell and explain.

ichim-david commented:
@dominiccooney I've addressed your suggestions in  https://github.com/sourcegraph/cody/pull/7022/commits/a48d316f7a5fc8deb5afa7e76da64d9c6264033b 
I will need help again with the updating of the agent recordings.

I really hope that as @kalanchan is prioritizing the changelog generator, a similar effort can be put into giving someone like myself a script or a bot command that I can run to update without needing extra permissions or software in order to avoid having to ask another dev todo this for me.

ichim-david commented:
@dominiccooney can you please update the agent recording if necessary? I would love to land this pull request, it has been stripped so much of what it used to be that now it simply removes the parenthesis as you've wanted, and it's a shame to have this change in limbo for longer.

abeatrix commented:
@ichim-david merged main and updated the recordings again.

I noticed now Cody would explain both the selection and the file instead of just the selected code after removing the `parentheses` . Would that cause unintended behavior? 

![image](https://github.com/user-attachments/assets/fde2fd50-1021-4793-a45d-d4d9aebbf17f)
